akka-stream/src/main/scala/akka/stream/impl/AcknowledgePublisher.scala

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import akka.actor.Props
+import akka.stream.OverflowStrategy
+import akka.stream.OverflowStrategy._
+
+/**
+ * INTERNAL API
+ */
+private[akka] object AcknowledgePublisher {
+  def props(bufferSize: Int, overflowStrategy: OverflowStrategy) =
+    Props(new AcknowledgePublisher(bufferSize, overflowStrategy))
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] class AcknowledgePublisher(bufferSize: Int, overflowStrategy: OverflowStrategy)
+  extends ActorRefSourceActor(bufferSize, overflowStrategy) {
+
+  override def receiveElem: Receive = {
+    case elem if isActive ⇒
+      if (totalDemand > 0L) {
+        onNext(elem)
+        sendAck(true)
+      } else if (bufferSize == 0) {
+        log.debug("Dropping element because there is no downstream demand: [{}]", elem)
+        sendAck(false)
+      } else if (!buffer.isFull)
+        enqueueAndSendAck(elem)
+      else overflowStrategy match {
+        case DropHead ⇒
+          buffer.dropHead()
+          enqueueAndSendAck(elem)
+        case DropTail ⇒
+          buffer.dropTail()
+          enqueueAndSendAck(elem)
+        case DropBuffer ⇒
+          buffer.clear()
+          enqueueAndSendAck(elem)
+        case DropNew ⇒
+          sendAck(false)
+        case Fail ⇒
+          onErrorThenStop(new Fail.BufferOverflowException(s"Buffer overflow (max capacity was: $bufferSize)!"))
+        case Backpressure ⇒
+        // there is a precondition check in Source.actorRefSource factory method
+      }
+  }
+
+  def enqueueAndSendAck(elem: Any): Unit = {
+    buffer.enqueue(elem)
+    sendAck(true)
+  }
+
+  def sendAck(ack: Boolean): Unit =
+    context.sender() ! ack
```

Comment:
I know it is an extra allocation to wrap the boolean, but we usually don't like naked types. Please introduce a wrapper message for that boolean.